Conflict docs (#92)
aerogear/offix: docs/ref-conflict-server.md

> # Server side conflict resolution

Reviewer: I think we should provide quick explanation as to exactly _what_ conflict resolution is in this context. Obvious for most but I think it could help based on what we discussed with some community members. Even just a sentence on the problem we are trying to solve would cover it.

> Offix provides out of the box conflict resolution for Node.JS platform.

Reviewer: ```suggestion
Offix provides out of the box conflict resolution for any Node.JS platform.
```

> For other languages you can follow Conflict specification chapter to implement custom detectiom mechanism for your resolvers.

Reviewer: ```suggestion
For other languages, you can follow the conflict specification chapter to implement a custom detection mechanism for your resolvers.
```

## Usage

> Install conflict package inside your Node.js project.

Reviewer: ```suggestion
Install the conflict package inside your Node.js project.
```

# Server Side Conflict Resolution

> When performing data synchronization between multiple clients it is common for remote devices to become offline for a certain amount of time or that data that was previously obtained by the client was modified on the server. As a result of being offline, data that is being modified by the client can become outdated with the server. Further operations on that record can cause conflict (often called a collision). 

Reviewer: ```suggestion
When performing data synchronization between multiple clients it is common for remote devices to become offline for a certain amount of time. As a result of being offline, data that is modified by a client can become outdated with the server. Further operations on that record can cause a conflict (often called a collision).
```

Resolvers can be implemented to handle conflicts on client .

> ## Implementing Custom Conflict Implementation

Reviewer: ```suggestion
## Implementing Custom Conflict Resolution
```

> The`ObjectState` interface is a complete conflict resolution implementation that provides a set of rules to detect and handle conflict. Interface will allow developers to handle conflict on the client. Client side application will need to match the server side implementation. Currently we support following implementations:

Reviewer: ```suggestion
The`ObjectState` interface is a complete conflict resolution implementation that provides a set of rules to detect and handle conflicts. The interface will allow developers to handle conflicts on the client. Client-side applications will need to match the server side implementation. Currently, we support the following implementations:
```

- `VersionObjectState` - allows to operate based on version field in schema
- `HashObjectState` - allows to operate based on object hashes

## Conflict Resolution Specification for other languages

> Conflict resolution package can be replicated to any language by implementing conflict detection logic in resolvers. 

Reviewer: ```suggestion
The conflict resolution package can be replicated to any language by implementing conflict detection logic in resolvers.
```

> Conflict detection should return specific format of error when conflict will be detected. This error will be processed by offix client and send back again as new mutation. To detect conflicts on the server 

Reviewer: ```suggestion
Conflict detection should return a specific format of error when a conflict has been detected. This error will be processed by offix-client and sent back again as a new mutation. To detect conflicts on the server 
```

> you can use one of many strategies for example:

Reviewer: ```suggestion
you can use one of many strategies, for example:
```

- Calculate hashes of the data
- Use specialized field like version or lastChangedDate
- Have separate storage with history of the changes

### Structure of the conflict error

> Server needs to return specific error when conflict is detected

Reviewer: ```suggestion
The server needs to return a specific error when a conflict is detected
```

> containing server and client states:

Reviewer: ```suggestion
containing both the server and client states:
```
